app/spotify.py

```diff
@@ -22,34 +65,65 @@ def get_songs_from_spotify_website(playlist):
     # raises SpotifyWebsiteParserException if parsing the website goes wrong
 
     return_data = []
-    if playlist.startswith("http"):
-        url = playlist.replace("spotify.com/playlist", "spotify.com/embed/playlist")
-    else:
-        url = base_url.format(playlist)
+    url_info = parse_uri(playlist)
 
-    req = requests.get(url)
+    req = requests.get(token_url)
     if req.status_code != 200:
-        raise SpotifyWebsiteParserException("ERROR: {} gave us not a 200. Instead: {}".format(url, req.status_code))
-
-    bs = BeautifulSoup(req.text, 'html.parser')
-    try:
-        songs_txt = bs.find('script', {'id': 'resource'}).string.strip()
-    except AttributeError:
-        raise SpotifyWebsiteParserException("ERROR: Could not get songs from Spotify website. Wrong Playlist id? Tried {}".format(url))
-    songs_json = json.loads(unquote(songs_txt))
-
-    for track in songs_json['tracks']['items']:
-        artist = track['track']['artists'][0]['name']
-        song = track['track']['name']
-        full = "{} {}".format(artist, song)
-        # remove everything in brackets to get better search results later on Deezer
-        # e.g. (Radio  Version) or (Remastered)
-        full = re.sub(r'\([^)]*\)', '', full)
-        return_data.append(full)
+        raise SpotifyWebsiteParserException(
+            "ERROR: {} gave us not a 200. Instead: {}".format(token_url, req.status_code))
+    token = req.json()
+
+    if url_info['type'] == "playlist":
+        url = playlist_base_url.format(url_info["id"])
+        while True:
+            resp = get_json_from_api(url, token["accessToken"])
+            for track in resp['items']:
+                return_data.append(parse_track(track["track"]))
+
+            if resp['next'] is None:
+                break
+            url = resp['next']
+    elif url_info["type"] == "track":
```

Comment:
Why do you make only a single retry for tracks/albums but infinity retries for a playlist download?

Reply:
The single retry is because of rate limiting. If we wait long enough (or specifically the number of seconds returned in the `Retry-After` is enough), then we should be able to receive the correct response. For playlists it's about retry-after header but more so it's about the pagination. In theory a playlist that has less than 100 items will be retried only once for the rate limiting just like tracks/albums.

Reply:
ah okay